README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ferriswheels <img src="man/figures/logo.gif" align="right" height="139" />

<!-- badges: start -->
<!-- badges: end -->

The goal of ferriswheels is to provide a fun harmless little data set to
play with

## Installation

You can install the development version of ferriswheels like so:

``` r
remotes::install_github("emilhvitfeldt/elevators")
```

## Example

This data set contains a number of interesting variables yo play with.

``` r
library(ferriswheels)
str(wheels)
#> Classes 'tbl_df', 'tbl' and 'data.frame':    74 obs. of  21 variables:
#>  $ name                 : chr  "360 Pensacola Beach" "Amuran" "Asiatique Sky" "Aurora Wheel" ...
#>  $ height               : num  200 303 200 295 180 ...
#>  $ diameter             : num  NA 200 200 272 NA ...
#>  $ opened               : Date, format: "2012-07-03" "2004-01-01" ...
#>  $ closed               : Date, format: "2013-01-01" NA ...
#>  $ country              : chr  "USA" "Japan" "Tailand" "Japan" ...
#>  $ location             : chr  "Pensacola Beach; Florida" "Kagoshima; Kyushu" "Asiatique the Riverfront" "Nagashima Spa Land; Mie; Honshu" ...
#>  $ number_of_cabins     : num  42 36 42 NA 40 48 42 NA 36 48 ...
#>  $ passengers_per_cabin : num  6 NA NA NA 6 40 8 NA 8 8 ...
#>  $ seating_capacity     : num  252 NA NA NA 240 1920 336 NA 288 384 ...
#>  $ hourly_capacity      : int  1260 NA NA NA 960 5760 1550 NA 1440 1152 ...
#>  $ ride_duration_minutes: num  12 14.5 NA NA 15 20 13 15 12 20 ...
#>  $ climate_controlled   : chr  "Yes" "Yes" "Yes" NA ...
#>  $ construction_cost    : chr  "Unknown" "Unknown" "Unknown" "Unknown" ...
#>  $ status               : chr  "Moved" "Operating" "Operating" "Operating" ...
#>  $ design_manufacturer  : chr  "Realty Masters of FL" NA "Dutch Wheels (Vekoma)" NA ...
#>  $ type                 : chr  "Transportable" NA NA "Fixed" ...
#>  $ vip_area             : chr  "Yes" NA NA NA ...
#>  $ ticket_cost_to_ride  : chr  NA NA NA NA ...
#>  $ official_website     : chr  NA NA "http://www.asiatiquesky.com/" "http://www.nagashima-onsen.co.jp/" ...
#>  $ turns                : num  4 1 NA NA NA 1 1 NA 3 NA ...
```

For starters, we can look at the diameter and height, colored by the
number of cabins

``` r
library(tidyverse)

wheels |>
  ggplot(aes(height, diameter, color = number_of_cabins)) +
  geom_point() +
  scale_color_viridis_b()
```

<img src="man/figures/README-unnamed-chunk-2-1.png" width="100%" />

If we look at the height of the Ferris wheels as a function of opening
date, we see that in the most recent time that the height have gotten
larger and larger.

``` r
wheels |>
  ggplot(aes(opened, height)) +
  geom_point()
```

<img src="man/figures/README-unnamed-chunk-3-1.png" width="100%" />

Now image we want to get a feel for how these Ferris wheels look. We can
construct them ourselves.

``` r
library(ggforce)

# We need complete data for these calculations
complete_wheels <- wheels |>
  select(name, height, diameter, number_of_cabins) |>
  drop_na()

set.seed(1234)
selected_complete_wheels <- complete_wheels |>
  # We have limited plotting space, so we are filtering to only include the
  # ferris wheels with the best spread out carts
  slice_max(order_by = diameter / number_of_cabins, n = 6, with_ties = FALSE)

# We are creating a seperate data.frame for the position of the carts
carts <- selected_complete_wheels |>
  group_by(name) |>
  summarise(
    cart = seq_len(number_of_cabins),
    # Get x and y for the carts
    cart_x = cos(cart / number_of_cabins * 2 * pi),
    cart_y = sin(cart / number_of_cabins * 2 * pi),
    # Size them to be the right distance from the center
    cart_x = cart_x * (diameter / 2),
    cart_y = cart_y * (diameter / 2),
    # Make sure the carts are raised enough
    cart_y = cart_y + height - diameter / 2,
    # Lower the carts just a bit so it appears they are hanging
    cart_y = cart_y - 12.5,
    cart_color = as.character(cart %% 3),
    .groups = "drop"
  )

selected_complete_wheels |>
  ggplot() +
  # Grass
  geom_abline(slope = 0, intercept = 0, color = "darkgreen") +
  ylim(0, NA) +
  # Ferris wheel circle
  geom_circle(aes(x0 = 0, y0 = height - diameter / 2, r = diameter / 2)) +
  # Left leg
  geom_segment(aes(x = -(height - diameter / 2)/2, xend = 0,
                   yend = height - diameter / 2, y = 0)) +
  # right leg
  geom_segment(aes(x = (height - diameter / 2)/2, xend = 0,
                   yend = height - diameter / 2, y = 0)) +
  # Carts
  geom_point(aes(cart_x, cart_y, fill = cart_color), data = carts, shape = 24) +
  facet_wrap(~name) +
  coord_fixed() +
  theme_minimal() +
  guides(fill = "none") +
  labs(y = NULL, x = NULL)
```

<img src="man/figures/README-unnamed-chunk-4-1.png" width="100%" />


NEWS.md
# ferriswheels 0.0.0.9000

* Added The Capital Wheel, National Harbor, Maryland


R/data.R
#' Ferris Wheels
#'
#' A data set containing information of the ferris wheels around the world.
#'
#' @format A data frame with `r nrow(wheels)` rows and `r ncol(wheels)`
#' variables:
#'
#' \describe{
#'   \item{name}{Character, name of ferris wheel}
#'   \item{height}{Double, height if ferris wheel in ft}
#'   \item{diameter}{Double, diameter of ferris wheel in ft}
#'   \item{opened}{Date, time ferris wheel was opened. January 1st was used for
#'     wheels where only the year was posted}
#'   \item{closed}{Date, time ferris wheel was closed. January 1st was used for
#'     wheels where only the year was posted}
#'   \item{country}{Character, country of origin}
#'   \item{location}{Character, location of ferris wheel. Sometimes city or
#'     theme park}
#'   \item{number_of_cabins}{Integer, number of cabins}
#'   \item{passengers_per_cabin}{Integer, Number of passengers per cabin}
#'   \item{seating_capacity}{Integer, total number of seats}
#'   \item{hourly_capacity}{Integer, number of passergers per hour}
#'   \item{ride_duration_minutes}{Integer, ride duration in minutes}
#'   \item{climate_controlled}{Character, are cabin climate controlled? Takes
#'     values `"Yes"`, `"no"` and `NA`}
#'   \item{construction_cost}{Character, approcimate construction cost}
#'   \item{status}{Character, status of ferris wheel}
#'   \item{design_manufacturer}{Character, name of manufacturer}
#'   \item{type}{Character, type of ferris wheel. One of "`Centerless`",
#'     "`Eccentric`", "`Enclosed`", "`Fixed`", "`Portable`", and
#'     "`Transportable`"}
#'   \item{vip_area}{Character, indictor of VIP area}
#'   \item{ticket_cost_to_ride}{Character, ticket price, not cleaned}
#'   \item{official_website}{Character, URL for ferris wheel}
#'   \item{turns}{Integer, number of turns per ride}
#' }
#'
#' @source \url{https://www.observationwheeldirectory.com/}
"wheels"


data-raw/wheels.R
## code to prepare `wheels` dataset goes here

library(tidyverse)
library(rvest)

wheels_urls <- read_html("https://www.observationwheeldirectory.com/wheels/") |>
  html_element("table") |>
  html_elements("a") |>
  html_attr("href")

extract_table <- function(x) {
  read_html(x) |>
    html_element("table") |>
    html_table()
}

all_tables <- map(wheels_urls, extract_table)

format_table <- function(x) {
  x <- x[, 1:2]
  x |>
    set_names(c("variable", "value")) |>
    mutate(name = names(x)[2])
}

all_tables_clean <- map_dfr(all_tables, format_table)

wheels <- all_tables_clean |>
  pivot_wider(names_from = variable, values_from = value)

wheels <- wheels |>
  janitor::clean_names() |>
  mutate(across(contains("turns"), as.integer)) |>
  mutate(turns = rowSums(cbind(turns, number_of_turns_rotations_per_ride,
           average_number_of_turns_rotations_per_ride), na.rm = TRUE)) |>
  mutate(turns = ifelse(turns == 0, NA, turns)) |>
  select(-number_of_turns_rotations_per_ride,
         -average_number_of_turns_rotations_per_ride) |>
  mutate(across(c(number_of_cabins, passengers_per_cabin,
                  seating_capacity, hourly_capacity, ride_duration_minutes),
                as.integer)) |>
  mutate(across(where(is.character), ~if_else(.x == "", NA_character_, .x)))

ft_calc <- function(x) {
  x |>
    str_extract("[\\d\\.]+ft") |>
    str_remove("ft") |>
    as.numeric()
}

date_calc <- function(x) {
  case_when(
    str_detect(x, "\\d+/\\d+/\\d+") ~ lubridate::mdy(x),
    str_detect(x, "\\d{4}") ~ lubridate::mdy(paste0("1/1/", x)),
    TRUE ~ lubridate::NA_Date_
  )
}

wheels <- wheels |>
  mutate(across(c(height, diameter), ft_calc)) |>
  mutate(across(c(opened, closed), date_calc))

wheels <- wheels |>
  mutate(ride_duration_minutes = rowSums(cbind(
    ride_duration_minutes,
    as.numeric(str_extract(ride_duration, "[\\d\\.]+"))
  ), na.rm = TRUE)) |>
  mutate(ride_duration_minutes = ifelse(ride_duration_minutes == 0, NA, ride_duration_minutes)) %>%
  select(-ride_duration)

# https://en.wikipedia.org/wiki/Capital_Wheel
wheels <- wheels |>
  add_row(
    name = "Capital Wheel",
    height = 175,
    diameter = 165,
    opened = lubridate::ymd("2014-05-23"),
    closed = NA,
    country = "USA",
    location = "National Harbor, Maryland",
    number_of_cabins = 42,
    passengers_per_cabin = 8,
    seating_capacity = 42 * 8,
    hourly_capacity = NA,
    ride_duration_minutes = 15,
    climate_controlled = "sometimes",
    construction_cost = NA,
    status = "Operating",
    design_manufacturer = "CWA Construction",
    type = NA,
    vip_area = "Yes",
    ticket_cost_to_ride = "$15 for adults, $13.50 for seniors, and $11.25 for children",
    official_website = "https://thecapitalwheel.com/",
    turns = NA
  )

usethis::use_data(wheels, overwrite = TRUE)
